src/validation.cpp

```diff
@@ -1663,6 +1672,65 @@ bool InitScriptExecutionCache(size_t max_size_bytes)
     return true;
 }
 
+struct DeferredCheckError
+{
+    std::string m_code;
+    std::string ToString() { return m_code; }
+};
+
+using UnvaultTriggerIdx = std::pair<unsigned int, std::optional<unsigned int>>;
+struct SimpleHash {
+    size_t operator()(const UnvaultTriggerIdx& p) const { return p.first ^ p.second.value_or(-1); }
+};
+
+
+//! Process all deferred script checks and perform the related validation against
+//! the transaction that generated them.
+static std::optional<DeferredCheckError> ValidateDeferredChecks(
+    const std::vector<DeferredCheck>& checks,
+    const CTransaction& tx)
+{
+    std::unordered_map<size_t, CAmount> expected_trigger_out;
+    std::unordered_map<size_t, CAmount> expected_recovery_out;
+
+    for (const auto& c : checks) {
+        if (const auto& recov_check = c.m_recov_spend_check) {
+            expected_recovery_out[recov_check->vout_idx] += recov_check->amount;
+        }
+        else if (const auto& trigger_check = c.m_vault_trigger_check) {
+            if (!trigger_check->revault_vout_idx) {
+                expected_trigger_out[trigger_check->vout_idx] += trigger_check->amount;
+            } else {
+                auto revault_amt{trigger_check->revault_amount};
+                auto trigger_amt{trigger_check->amount - revault_amt};
+                Assume(trigger_amt >= 0);
+
+                expected_trigger_out[trigger_check->vout_idx] += trigger_amt;
```

Comment:
nit (feel free to ignore): Add a comment why this cannot overflow.

This sum is bounded by total bitcoins. Hence, this sum be atmax 21 * 10**6 * 10**8.

Reply:
Added a comment.